src/main/scala/com/gu/identity/testing/usernames/TestUsernames.scala

```diff
@@ -1,63 +1,89 @@
+        logger.warn(s"TestUsername created on $creationTime, apparently AFTER this moment in time! now=$now")
+      }
 
-  val ConfPayloadByteSize = 2
+      val confPayload: Array[Byte] = Array.ofDim[Byte](ConfPayloadByteSize)
+      bb.get(confPayload)
 
-  def apply(usernameEncoder: Encoder, recency: Duration = ofMinutes(30))(implicit clock: Clock = systemUTC): TestUsernames = new TestUsernames {
-    def generate(conf: Array[Byte]): String = {
-      val bb = ByteBuffer.allocate(Encoder.PayloadByteLength)
-      bb.putInt(Instant.now(clock).getEpochSecond.toInt)
-      bb.put(conf)
-      usernameEncoder.encodeSigned(bb.array())
+      val expirationTime = creationTime.plus(recency)
+      if (expirationTime > now) Some(confPayload) else {
+        logger.warn(s"TestUsername created on $creationTime EXPIRED on $expirationTime")
+        None
+      }
     }
 
-    def validate(username: String): Option[Array[Byte]] = {
+    for {
+      validlySignedData <- usernameEncoder.decodeSigned(username)
+      recentlyCreatedConfData <- extractRecentConfData(validlySignedData)
+    } yield recentlyCreatedConfData
+  }
+
+  def isValid(username: String): Boolean = validate(username).isDefined
+
+  def generateEmail(baseEmail: String): TestUser = {
+    val token = generate(Array.ofDim[Byte](2))
+    val email: String = tokenToEmail(baseEmail, token)
+    TestUser(email, token)
+  }
 
-      import scala.math.Ordering.Implicits._
+  def isValidEmail(primaryEmailAddress: String): Boolean = {
+    val maybeValidTestUser = maybeTokenFromEmail(primaryEmailAddress)
+    maybeValidTestUser.exists(isValid)
+  }
 
-      def extractRecentConfData(validlySignedData: Array[Byte]): Option[Array[Byte]] = {
-        val bb = ByteBuffer.wrap(validlySignedData)
-        val creationTime = ofEpochSecond(bb.getInt())
-        val now = Instant.now(clock)
-        if (creationTime > now) {
-          logger.warn(s"TestUsername created on $creationTime, apparently AFTER this moment in time! now=$now")
-        }
+}
 
-        val confPayload: Array[Byte] = Array.ofDim[Byte](ConfPayloadByteSize)
-        bb.get(confPayload)
+object TestUsernames extends LazyLogging {
+
+  case class TestUser(email: String, token: String)
 
-        val expirationTime = creationTime.plus(recency)
-        if (expirationTime > now) Some(confPayload) else {
-          logger.warn(s"TestUsername created on $creationTime EXPIRED on $expirationTime")
-          None
-        }
+  def maybeTokenFromEmail(primaryEmailAddress: String): Option[String] =
+    for {
+      localPart <- primaryEmailAddress.split('@').toList match {
+        case local :: _ :: Nil => Some(local)
+        case _ => None
+      }
+      possibleTestUsername <- localPart.split('+').toList match {
+        case _ :: subAddress :: _ => Some(subAddress)
+        case _ => None
       }
+    } yield possibleTestUsername
+
+  def tokenToEmail(baseEmail: String, token: String) =
+    baseEmail.replace("@", s"+$token@")
```

Comment:
these are the new functions that insert and extract the subaddress from the email address